cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/AppFabricServer.java

```diff
@@ -181,6 +198,70 @@ protected void startUp() throws Exception {
                                                                       applicationCount);
     metricsCollectionService.getContext(Collections.emptyMap()).gauge(Constants.Metrics.Program.NAMESPACE_COUNT,
                                                                       namespaceCount);
+    // Fetch all apps to emit metadata for plugins
+    // This is used to back-fill plugin metadata after update
+    Map<ApplicationId, ApplicationMeta> apps =
```

Comment:
we shouldn't be doing expensive operations like this in the startUp() method. Startup should start a thread or service that will begin this work, but it shouldn't actually do all the reading here.

Reply:
this is also a lot of work to try to do in a single transaction. When running on Hadoop/standalone, this can timeout and fail.

It would be better to break it up, at least namespace by namespace.